File: ch07/sec7.4/events/signal/sig.py
import signal
import asyncio
import os
from typing import Callable, Dict, List

# Interrupt Vector Table (IVT)
class InterruptDispatcher:
    def __init__(self):
        self.table: Dict[int, List[Callable[[], None]]] = {}

    def register(self, sig: int, handler: Callable[[], None]):
        if sig not in self.table:
            self.table[sig] = []
        self.table[sig].append(handler)

    def dispatch(self, sig: int):
        print(f"[IVT] Signal {sig} received. Dispatching...")
        for handler in self.table.get(sig, []):
            asyncio.ensure_future(self._run_async(handler))

    async def _run_async(self, handler: Callable[[], None]):
        await asyncio.sleep(0)  # Yield to event loop
        handler()

# Global dispatcher instance
dispatcher = InterruptDispatcher()

# Connect dispatcher to OS-level signal
def signal_handler(sig_num, frame):
    dispatcher.dispatch(sig_num)

# Register signal handler for SIGUSR1
signal.signal(signal.SIGUSR1, signal_handler)

# Example async-safe observer handlers
def handler_A():
    print("Handler A reacting to signal!")

def handler_B():
    print("Handler B doing something else!")

# Register observers to SIGUSR1
dispatcher.register(signal.SIGUSR1, handler_A)
dispatcher.register(signal.SIGUSR1, handler_B)

# Run a simple event loop to await signals
async def main():
    print(f"Send SIGUSR1 using: kill -USR1 {os.getpid()}")
    while True:
        await asyncio.sleep(1)

if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("Shutting down.")

File: ch07/sec7.4/signal/sig.c
#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <unistd.h>

#define MAX_HANDLERS 10
#define MAX_SIGNALS  32

// Function pointer type for handlers
typedef void (*Handler)();

// Interrupt Vector Table: array of handler lists per signal
typedef struct {
    Handler handlers[MAX_HANDLERS];
    int count;
} SignalHandlerList;

SignalHandlerList ivt[MAX_SIGNALS];

// Register a handler for a signal
void register_handler(int sig, Handler h) {
    if (sig >= MAX_SIGNALS) return;
    if (ivt[sig].count < MAX_HANDLERS) {
        ivt[sig].handlers[ivt[sig].count++] = h;
    }
}

// Dispatcher called from signal handler
void dispatch(int sig) {
    printf("[IVT] Signal %d received. Dispatching...\n", sig);
    for (int i = 0; i < ivt[sig].count; i++) {
        if (ivt[sig].handlers[i]) {
            ivt[sig].handlers[i]();
        }
    }
}

// Real signal handler installed via sigaction
void signal_handler(int sig) {
    dispatch(sig);
}

// Example handlers
void handler_A() {
    printf("Handler A responding to signal!\n");
}

void handler_B() {
    printf("Handler B taking action!\n");
}

int main() {
    printf("PID: %d\n", getpid());
    printf("Send SIGUSR1 using: kill -USR1 %d\n", getpid());

    // Register signal and handlers
    struct sigaction sa;
    sa.sa_handler = signal_handler;
    sa.sa_flags = SA_RESTART;
    sigemptyset(&sa.sa_mask);
    sigaction(SIGUSR1, &sa, NULL);

    register_handler(SIGUSR1, handler_A);
    register_handler(SIGUSR1, handler_B);

    // Run forever
    while (1) {
        pause();  // Wait for signal
    }

    return 0;
}